// twitter.js/dist/util/Constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamType = exports.defaultClientOptions = exports.ClientEvents = void 0;
const twitter_types_1 = require("twitter-types");
exports.ClientEvents = {
    FILTERED_TWEET_CREATE: 'filteredTweetCreate',
    PARTIAL_ERROR: 'partialError',
    READY: 'ready',
    SAMPLED_TWEET_CREATE: 'sampledTweetCreate',
};
exports.defaultClientOptions = {
    api: {
        version: 2,
        baseURL: 'https://api.twitter.com',
    },
    queryParameters: {
        userFields: twitter_types_1.APIUserFieldsParameters,
        tweetFields: twitter_types_1.APITweetFieldsParameters.filter(f => !['promoted_metrics', 'organic_metrics', 'non_public_metrics'].includes(f)),
        spaceFields: twitter_types_1.APISpaceFieldsParameters.filter(f => f !== 'subscriber_count'),
        mediaFields: twitter_types_1.APIMediaFieldsParameters,
        placeFields: twitter_types_1.APIPlaceFieldsParameters,
        pollFields: twitter_types_1.APIPollFieldsParameters,
        listFields: twitter_types_1.APIListFieldsParameters,
        tweetExpansions: twitter_types_1.APITweetExpansionsParameters,
        userExpansions: twitter_types_1.APIUserExpansionsParameters,
        spaceExpansions: twitter_types_1.APISpaceExpansionsParameters,
        listExpansions: twitter_types_1.APIListExpansionsParameters,
    },
    events: [],
};
exports.StreamType = {
    SAMPLED: 'sampled',
    FILTERED: 'filtered',
};
