// twitter.js/dist/structures/Media.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Media = void 0;
const misc_1 = require("./misc");
class Media {
    constructor(client, data) {
        Object.defineProperty(this, 'client', { writable: true, enumerable: false });
        this.client = client;
        this.id = data.media_key;
        this.type = data.type;
        this.duration = data.duration_ms ?? null;
        this.height = data.height ?? null;
        this.nonPublicMetrics = data.non_public_metrics ? new misc_1.MediaNonPublicMetrics(data.non_public_metrics) : null;
        this.organicMetrics = data.organic_metrics ? new misc_1.MediaOrganicMetrics(data.organic_metrics) : null;
        this.previewImageURL = data.preview_image_url ?? null;
        this.promotedMetrics = data.promoted_metrics ? new misc_1.MediaPromotedMetrics(data.promoted_metrics) : null;
        this.publicMetrics = data.public_metrics ? new misc_1.MediaPublicMetrics(data.public_metrics) : null;
        this.url = data.url ?? null;
        this.width = data.width ?? null;
    }
}
exports.Media = Media;
