// twitter.js/dist/structures/SimplifiedSpace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimplifiedSpace = void 0;
const BaseStructure_1 = require("./BaseStructure");
class SimplifiedSpace extends BaseStructure_1.BaseStructure {
    constructor(client, data) {
        super(client, data);
        this.state = data.state;
        this.createdAt = data.created_at ? new Date(data.created_at) : null;
        this.creatorId = data.creator_id ?? null;
        this.hostIds = data.host_ids ?? [];
        this.invitedUserIds = data.invited_user_ids ?? [];
        this.isTicketed = data.is_ticketed ?? null;
        this.lang = data.lang ?? null;
        this.participantCount = data.participant_count ?? null;
        this.scheduledStart = data.scheduled_start ? new Date(data.scheduled_start) : null;
        this.speakerIds = data.speaker_ids ?? [];
        this.startedAt = data.started_at ? new Date(data.started_at) : null;
        this.title = data.title ?? null;
        this.updatedAt = data.updated_at ? new Date(data.updated_at) : null;
        this.subscriberCount = data.subscriber_count ?? null;
    }
    async fetchSharedTweets(options) {
        return this.client.spaces.fetchSharedTweets(this.id, options);
    }
}
exports.SimplifiedSpace = SimplifiedSpace;
