// twitter.js/dist/structures/SimplifiedTweet.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SimplifiedTweet_instances, _SimplifiedTweet_patchReferencedTweets, _SimplifiedTweet_patchTweetContextAnnotations;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimplifiedTweet = void 0;
const BaseStructure_1 = require("./BaseStructure");
const misc_1 = require("./misc");
class SimplifiedTweet extends BaseStructure_1.BaseStructure {
    constructor(client, data) {
        super(client, data);
        _SimplifiedTweet_instances.add(this);
        this.text = data.text;
        this.attachments = data.attachments ? new misc_1.TweetAttachments(data.attachments) : null;
        this.authorId = data.author_id ?? null;
        this.contextAnnotations = __classPrivateFieldGet(this, _SimplifiedTweet_instances, "m", _SimplifiedTweet_patchTweetContextAnnotations).call(this, data.context_annotations);
        this.conversationId = data.conversation_id ?? null;
        this.createdAt = data.created_at ? new Date(data.created_at) : null;
        this.entities = data.entities ? new misc_1.TweetEntities(data.entities) : null;
        this.geo = data.geo ? new misc_1.TweetGeo(data.geo) : null;
        this.inReplyToUserId = data.in_reply_to_user_id ?? null;
        this.lang = data.lang ?? null;
        this.nonPublicMetrics = data.non_public_metrics ? new misc_1.TweetNonPublicMetrics(data.non_public_metrics) : null;
        this.organicMetrics = data.organic_metrics ? new misc_1.TweetOrganicMetrics(data.organic_metrics) : null;
        this.possiblySensitive = data.possibly_sensitive ?? null;
        this.promotedMetrics = data.promoted_metrics ? new misc_1.TweetPromotedMetrics(data.promoted_metrics) : null;
        this.publicMetrics = data.public_metrics ? new misc_1.TweetPublicMetrics(data.public_metrics) : null;
        this.referencedTweets = __classPrivateFieldGet(this, _SimplifiedTweet_instances, "m", _SimplifiedTweet_patchReferencedTweets).call(this, data.referenced_tweets);
        this.replySettings = data.reply_settings ?? null;
        this.source = data.source ?? null;
        this.withheld = data.withheld ? new misc_1.TweetWithheld(data.withheld) : null;
    }
    async like() {
        return this.client.tweets.like(this.id);
    }
    async unlike() {
        return this.client.tweets.unlike(this.id);
    }
    async hide() {
        return this.client.tweets.hide(this.id);
    }
    async unhide() {
        return this.client.tweets.unhide(this.id);
    }
    async retweet() {
        return this.client.tweets.retweet(this.id);
    }
    async unRetweet() {
        return this.client.tweets.unRetweet(this.id);
    }
    async reply(options) {
        return this.client.tweets.create({ ...options, inReplyToTweet: this.id });
    }
    async quote(options) {
        return this.client.tweets.create({ ...options, quoteTweet: this.id });
    }
    async delete() {
        return this.client.tweets.delete(this.id);
    }
}
exports.SimplifiedTweet = SimplifiedTweet;
_SimplifiedTweet_instances = new WeakSet(), _SimplifiedTweet_patchReferencedTweets = function _SimplifiedTweet_patchReferencedTweets(rawTweetReferences) {
    const tweetReferencesArray = [];
    if (!rawTweetReferences)
        return tweetReferencesArray;
    for (const rawTweetReference of rawTweetReferences) {
        const tweetReference = new misc_1.TweetReference(rawTweetReference);
        tweetReferencesArray.push(tweetReference);
    }
    return tweetReferencesArray;
}, _SimplifiedTweet_patchTweetContextAnnotations = function _SimplifiedTweet_patchTweetContextAnnotations(rawContextAnnotations) {
    const tweetContextAnnotationsArray = [];
    if (!rawContextAnnotations)
        return tweetContextAnnotationsArray;
    for (const rawContextAnnotation of rawContextAnnotations) {
        const contextAnnotation = new misc_1.TweetContextAnnotation(rawContextAnnotation);
        tweetContextAnnotationsArray.push(contextAnnotation);
    }
    return tweetContextAnnotationsArray;
};
