// twitter.js/dist/structures/SimplifiedUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimplifiedUser = void 0;
const BaseStructure_1 = require("./BaseStructure");
const misc_1 = require("./misc");
class SimplifiedUser extends BaseStructure_1.BaseStructure {
    constructor(client, data) {
        super(client, data);
        this.name = data.name;
        this.username = data.username;
        this.createdAt = data.created_at ? new Date(data.created_at) : null;
        this.description =
            typeof data.description === 'string' ? (data.description.length > 0 ? data.description : null) : null;
        this.entities = data.entities ? new misc_1.UserEntities(data.entities) : null;
        this.location = data.location ?? null;
        this.pinnedTweetId = data.pinned_tweet_id ?? null;
        this.profileImageURL = data.profile_image_url ?? null;
        this.protected = data.protected ?? null;
        this.publicMetrics = data.public_metrics ? new misc_1.UserPublicMetrics(data.public_metrics) : null;
        this.url = data.url ?? null;
        this.verified = data.verified ?? null;
        this.withheld = data.withheld ? new misc_1.UserWitheld(data.withheld) : null;
    }
    async follow() {
        return this.client.users.follow(this.id);
    }
    async unfollow() {
        return this.client.users.unfollow(this.id);
    }
    async block() {
        return this.client.users.block(this.id);
    }
    async unblock() {
        return this.client.users.unblock(this.id);
    }
    async mute() {
        return this.client.users.mute(this.id);
    }
    async unmute() {
        return this.client.users.unmute(this.id);
    }
}
exports.SimplifiedUser = SimplifiedUser;
