// twitter.js/dist/structures/SimplifiedList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SimplifiedList = void 0;
const BaseStructure_1 = require("./BaseStructure");
class SimplifiedList extends BaseStructure_1.BaseStructure {
    constructor(client, data) {
        super(client, data);
        this.name = data.name;
        this.description = data.description ?? null;
        this.private = data.private ?? null;
        this.followerCount = data.follower_count ?? null;
        this.memberCount = data.member_count ?? null;
        this.ownerId = data.owner_id ?? null;
        this.createdAt = data.created_at ?? null;
    }
    async delete() {
        return this.client.lists.delete(this.id);
    }
    async update(options) {
        return this.client.lists.update(this.id, options);
    }
    async addMember(user) {
        return this.client.lists.addMember(this.id, user);
    }
    async removeMember(user) {
        return this.client.lists.removeMember(this.id, user);
    }
    async follow() {
        return this.client.lists.follow(this.id);
    }
    async unfollow() {
        return this.client.lists.unfollow(this.id);
    }
    async pin() {
        return this.client.lists.pin(this.id);
    }
    async unpin() {
        return this.client.lists.unpin(this.id);
    }
    async setName(name) {
        return this.update({ name });
    }
    async setDescription(description) {
        return this.update({ description });
    }
    async setPrivate(isPrivate) {
        return this.update({ private: isPrivate });
    }
}
exports.SimplifiedList = SimplifiedList;
