// twitter.js/dist/structures/misc/MediaMetrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MediaPromotedMetrics = exports.MediaOrganicMetrics = exports.MediaPublicNonPublicMetricsUnion = exports.MediaPublicMetrics = exports.MediaNonPublicMetrics = void 0;
class MediaNonPublicMetrics {
    constructor(data) {
        this.playbackCountZero = data.playback_0_count;
        this.playbackCountTwentyFive = data.playback_25_count;
        this.playbackCountFifty = data.playback_50_count;
        this.playbackCountSeventyFive = data.playback_75_count;
        this.playbackCountHundred = data.playback_100_count;
    }
}
exports.MediaNonPublicMetrics = MediaNonPublicMetrics;
class MediaPublicMetrics {
    constructor(data) {
        this.viewCount = data.view_count;
    }
}
exports.MediaPublicMetrics = MediaPublicMetrics;
class MediaPublicNonPublicMetricsUnion extends MediaNonPublicMetrics {
    constructor(data) {
        super(data);
        this.viewCount = data.view_count;
    }
}
exports.MediaPublicNonPublicMetricsUnion = MediaPublicNonPublicMetricsUnion;
class MediaOrganicMetrics extends MediaPublicNonPublicMetricsUnion {
    constructor(data) {
        super(data);
    }
}
exports.MediaOrganicMetrics = MediaOrganicMetrics;
class MediaPromotedMetrics extends MediaPublicNonPublicMetricsUnion {
    constructor(data) {
        super(data);
    }
}
exports.MediaPromotedMetrics = MediaPromotedMetrics;
