// twitter.js/dist/structures/misc/TweetMetrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TweetPublicMetrics = exports.TweetPromotedMetrics = exports.TweetOrganicMetrics = exports.TweetNonPublicMetrics = void 0;
class TweetNonPublicMetrics {
    constructor(data) {
        this.impressionCount = data.impression_count;
        this.urlLinkClicks = data.url_link_clicks ?? null;
        this.userProfileClicks = data.user_profile_clicks;
    }
}
exports.TweetNonPublicMetrics = TweetNonPublicMetrics;
class TweetOrganicMetrics extends TweetNonPublicMetrics {
    constructor(data) {
        super(data);
        this.likeCount = data.like_count;
        this.replyCount = data.reply_count;
        this.retweetCount = data.retweet_count;
    }
}
exports.TweetOrganicMetrics = TweetOrganicMetrics;
class TweetPromotedMetrics extends TweetOrganicMetrics {
    constructor(data) {
        super(data);
    }
}
exports.TweetPromotedMetrics = TweetPromotedMetrics;
class TweetPublicMetrics {
    constructor(data) {
        this.retweetCount = data.retweet_count;
        this.replyCount = data.reply_count;
        this.likeCount = data.like_count;
        this.quoteCount = data.quote_count;
    }
}
exports.TweetPublicMetrics = TweetPublicMetrics;
