// twitter.js/dist/books/ListFollowersBook.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _ListFollowersBook_instances, _ListFollowersBook_fetchPages;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ListFollowersBook = void 0;
const util_1 = require("../util");
const BaseBook_1 = require("./BaseBook");
const errors_1 = require("../errors");
const structures_1 = require("../structures");
class ListFollowersBook extends BaseBook_1.BaseBook {
    constructor(client, options) {
        super(client, options);
        _ListFollowersBook_instances.add(this);
        const listId = client.lists.resolveId(options.list);
        if (!listId)
            throw new errors_1.CustomError('LIST_RESOLVE_ID', 'create ListFollowersBook for');
        this.listId = listId;
    }
    async fetchNextPage() {
        if (!this._hasMadeInitialRequest) {
            this._hasMadeInitialRequest = true;
            return __classPrivateFieldGet(this, _ListFollowersBook_instances, "m", _ListFollowersBook_fetchPages).call(this);
        }
        if (!this._nextToken)
            throw new errors_1.CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _ListFollowersBook_instances, "m", _ListFollowersBook_fetchPages).call(this, this._nextToken);
    }
    async fetchPreviousPage() {
        if (!this._previousToken)
            throw new errors_1.CustomError('PAGINATED_RESPONSE_HEAD_REACHED');
        return __classPrivateFieldGet(this, _ListFollowersBook_instances, "m", _ListFollowersBook_fetchPages).call(this, this._previousToken);
    }
}
exports.ListFollowersBook = ListFollowersBook;
_ListFollowersBook_instances = new WeakSet(), _ListFollowersBook_fetchPages = async function _ListFollowersBook_fetchPages(token) {
    const listFollowers = new util_1.Collection();
    const queryParameters = this.client.options.queryParameters;
    const query = {
        expansions: queryParameters?.userExpansions,
        'tweet.fields': queryParameters?.tweetFields,
        'user.fields': queryParameters?.userFields,
        pagination_token: token,
    };
    if (this.maxResultsPerPage)
        query.max_results = this.maxResultsPerPage;
    const requestData = new structures_1.RequestData({ query });
    const data = await this.client._api.users(this.listId).followers.get(requestData);
    this._nextToken = data.meta.next_token;
    this._previousToken = data.meta.previous_token;
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.result_count === 0)
        return listFollowers;
    const rawUsers = data.data;
    const rawIncludes = data.includes;
    for (const rawUser of rawUsers) {
        const user = this.client.users._add(rawUser.id, { data: rawUser, includes: rawIncludes }, false);
        listFollowers.set(user.id, user);
    }
    return listFollowers;
};
