// twitter.js/dist/books/MemberOfListsBook.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _MemberOfListsBook_instances, _MemberOfListsBook_fetchPages;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MemberOfListsBook = void 0;
const util_1 = require("../util");
const BaseBook_1 = require("./BaseBook");
const errors_1 = require("../errors");
const structures_1 = require("../structures");
class MemberOfListsBook extends BaseBook_1.BaseBook {
    constructor(client, options) {
        super(client, options);
        _MemberOfListsBook_instances.add(this);
        const userId = client.users.resolveId(options.user);
        if (!userId)
            throw new errors_1.CustomError('USER_RESOLVE_ID', 'create MemberOfListsBook for');
        this.userId = userId;
    }
    async fetchNextPage() {
        if (!this._hasMadeInitialRequest) {
            this._hasMadeInitialRequest = true;
            return __classPrivateFieldGet(this, _MemberOfListsBook_instances, "m", _MemberOfListsBook_fetchPages).call(this);
        }
        if (!this._nextToken)
            throw new errors_1.CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _MemberOfListsBook_instances, "m", _MemberOfListsBook_fetchPages).call(this, this._nextToken);
    }
    async fetchPreviousPage() {
        if (!this._previousToken)
            throw new errors_1.CustomError('PAGINATED_RESPONSE_HEAD_REACHED');
        return __classPrivateFieldGet(this, _MemberOfListsBook_instances, "m", _MemberOfListsBook_fetchPages).call(this, this._previousToken);
    }
}
exports.MemberOfListsBook = MemberOfListsBook;
_MemberOfListsBook_instances = new WeakSet(), _MemberOfListsBook_fetchPages = async function _MemberOfListsBook_fetchPages(token) {
    const lists = new util_1.Collection();
    const queryParameters = this.client.options.queryParameters;
    const query = {
        expansions: queryParameters?.listExpansions,
        'user.fields': queryParameters?.userFields,
        'list.fields': queryParameters?.listFields,
        pagination_token: token,
    };
    if (this.maxResultsPerPage)
        query.max_results = this.maxResultsPerPage;
    const requestData = new structures_1.RequestData({ query });
    const data = await this.client._api
        .users(this.userId)
        .list_memberships.get(requestData);
    this._nextToken = data.meta.next_token;
    this._previousToken = data.meta.previous_token;
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.result_count === 0)
        return lists;
    const rawLists = data.data;
    const rawIncludes = data.includes;
    for (const rawList of rawLists) {
        const list = this.client.lists._add(rawList.id, { data: rawList, includes: rawIncludes }, false);
        lists.set(list.id, list);
    }
    return lists;
};
