// twitter.js/dist/books/SearchTweetsBook.js
"use strict";
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _SearchTweetsBook_instances, _SearchTweetsBook_fetchPages;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchTweetsBook = void 0;
const util_1 = require("../util");
const errors_1 = require("../errors");
const structures_1 = require("../structures");
const BaseRangeBook_1 = require("./BaseRangeBook");
class SearchTweetsBook extends BaseRangeBook_1.BaseRangeBook {
    constructor(client, options) {
        super(client, options);
        _SearchTweetsBook_instances.add(this);
        this.query = options.query;
        this.sortOrder = options.sortOrder ?? null;
    }
    async fetchNextPage() {
        if (!this._hasMadeInitialRequest) {
            this._hasMadeInitialRequest = true;
            return __classPrivateFieldGet(this, _SearchTweetsBook_instances, "m", _SearchTweetsBook_fetchPages).call(this);
        }
        if (!this._nextToken)
            throw new errors_1.CustomError('PAGINATED_RESPONSE_TAIL_REACHED');
        return __classPrivateFieldGet(this, _SearchTweetsBook_instances, "m", _SearchTweetsBook_fetchPages).call(this, this._nextToken);
    }
}
exports.SearchTweetsBook = SearchTweetsBook;
_SearchTweetsBook_instances = new WeakSet(), _SearchTweetsBook_fetchPages = async function _SearchTweetsBook_fetchPages(token) {
    const tweets = new util_1.Collection();
    const queryParameters = this.client.options.queryParameters;
    const query = {
        expansions: queryParameters?.tweetExpansions,
        'tweet.fields': queryParameters?.tweetFields,
        'user.fields': queryParameters?.userFields,
        'media.fields': queryParameters?.mediaFields,
        'place.fields': queryParameters?.placeFields,
        'poll.fields': queryParameters?.pollFields,
        query: this.query,
        next_token: token,
        sort_order: this.sortOrder ?? undefined,
        since_id: this.afterTweetId ?? undefined,
        until_id: this.beforeTweetId ?? undefined,
        max_results: this.maxResultsPerPage ?? undefined,
        start_time: this.startTimestamp ? new Date(this.startTimestamp).toISOString() : undefined,
        end_time: this.endTimestamp ? new Date(this.endTimestamp).toISOString() : undefined,
    };
    const requestData = new structures_1.RequestData({ query });
    const data = await this.client._api.tweets.search.recent.get(requestData);
    this._nextToken = data.meta.next_token;
    this.hasMore = data.meta.next_token ? true : false;
    if (data.meta.result_count === 0)
        return tweets;
    const rawTweets = data.data;
    const rawIncludes = data.includes;
    for (const rawTweet of rawTweets) {
        const tweet = this.client.tweets._add(rawTweet.id, { data: rawTweet, includes: rawIncludes }, false);
        tweets.set(tweet.id, tweet);
    }
    return tweets;
};
